File: src/main/java/com/yunseong/netty/NettyApplication.java
package com.yunseong.netty;

import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.*;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.handler.codec.http.HttpServerCodec;
import io.netty.handler.logging.LogLevel;
import io.netty.handler.logging.LoggingHandler;

public class NettyApplication {

    public static void main(String[] args) throws InterruptedException {
        EventLoopGroup bossGroup = (new NioEventLoopGroup());
        EventLoopGroup workerGroup = (new NioEventLoopGroup());
        ServerBootstrap b = new ServerBootstrap();
        b.group(bossGroup, workerGroup).channel(NioServerSocketChannel.class)
            .handler((new LoggingHandler(LogLevel.INFO)))
            .childHandler((new ChannelInitializer<>() {
                @Override
                protected void initChannel(Channel ch) {
                    ChannelPipeline p = ch.pipeline();
                    p.addLast((new HttpServerCodec()));
                    p.addLast((new HttpServerInboundHandler()));
                }
        }));
        ChannelFuture future = b.bind(Integer.parseInt(System.getProperty("port", "8080"))).sync();
        future.channel().closeFuture().sync();
    }
}

File: src/main/kotlin/com/yunseong/netty/NettyApplication.kt
package com.yunseong.netty

import io.netty.bootstrap.ServerBootstrap
import io.netty.buffer.ByteBuf
import io.netty.channel.*
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioServerSocketChannel
import io.netty.handler.codec.http.*
import io.netty.handler.logging.LogLevel
import io.netty.handler.logging.LoggingHandler
import java.nio.charset.Charset

fun main() {
    val bossGroup: EventLoopGroup = NioEventLoopGroup()
    val workerGroup: EventLoopGroup = NioEventLoopGroup()
    val b = ServerBootstrap()
    b.group(bossGroup, workerGroup)
        .channel(NioServerSocketChannel::class.java)
        .handler(LoggingHandler(LogLevel.INFO))
        .childHandler(HttpServerChannelInitializer())
    val future = b.bind(System.getProperty("port", "8080").toInt()).sync()
    future.channel().closeFuture().sync()
}

internal class DiscardServerHandler : SimpleChannelInboundHandler<Any?>() {

    @Throws(Exception::class)
    override fun channelRead0(ctx: ChannelHandlerContext, msg: Any?) {
    }

    @Throws(Exception::class)
    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        cause.printStackTrace()
        ctx.close()
    }
}

internal class EchoServerHandler : ChannelInboundHandlerAdapter() {

    override fun channelRead(ctx: ChannelHandlerContext?, msg: Any?) {
        val readMessage: String = (msg as ByteBuf).toString(Charset.defaultCharset())

        println("수신한 문자열 [$readMessage]")

        ctx?.write(msg)
    }

    override fun channelReadComplete(ctx: ChannelHandlerContext?) {
        ctx?.flush()
    }
}